src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -21,8 +30,104 @@ public function getUrl(string $bucket, string $key): string
+            $this->putObject(array_merge($options, ['Bucket' => $bucket, 'Key' => $key, 'Body' => $object]));
+
+            return;
+        }
+
+        $parts = [];
+        $partNumber = 1;
+        $uploadId = $this->createMultipartUpload(array_merge($options, ['Bucket' => $bucket, 'Key' => $key]))->getUploadId();
+        foreach ($stream as $chunk) {
+            if (1 === $partNumber && \strlen($chunk) < 5 * $megabyte) {
+                /*
+                 * The first part is too small to upload using MultipartUpload.
+                 * AWS has a limit of minimum 5MB.
+                 * Because we are using FixedSizeStream, we know that this is the
+                 * last chunk or we would had have a full "chunk size".
+                 *
+                 * Lets abort multipart upload and use a normal upload.
+                 */
+                $this->abortMultipartUpload(['Bucket' => $bucket, 'Key' => $key, 'UploadId' => $uploadId]);
+                $this->putObject(array_merge($options, ['Bucket' => $bucket, 'Key' => $key, 'Body' => $chunk]));
+
+                return;
+            }
+
+            try {
+                $response = $this->uploadPart([
+                    'Bucket' => $bucket,
+                    'Key' => $key,
+                    'UploadId' => $uploadId,
+                    'PartNumber' => $partNumber,
+                    'Body' => $chunk,
+                ]);
+                $parts[] = new CompletedPart(['ETag' => $response->getETag(), 'PartNumber' => $partNumber]);
+                ++$partNumber;
+            } catch (\Throwable $e) {
+                $this->abortMultipartUpload(['Bucket' => $bucket, 'Key' => $key, 'UploadId' => $uploadId]);
+
+                throw $e;
+            }
+        }
+
+        if (empty($parts)) {
+            // The upload did not contain any data.
+            $this->abortMultipartUpload(['Bucket' => $bucket, 'Key' => $key, 'UploadId' => $uploadId]);
+
+            return;
+        }
+
+        $this->completeMultipartUpload([
+            'Bucket' => $bucket,
+            'Key' => $key,
+            'UploadId' => $uploadId,
+            'MultipartUpload' => new CompletedMultipartUpload(['Parts' => $parts]),
+        ]);
+    }
+
+    private function getStream($object, int $chunkSize): FixedSizeStream
+    {
+        if (null === $object || \is_string($object)) {
+            $stream = StringStream::create($object ?? '');
+        } elseif (\is_resource($object)) {
+            $stream = ResourceStream::create($object, $chunkSize);
+        } elseif (\is_callable($object)) {
+            $stream = CallableStream::create($object, $chunkSize);
+        } elseif (\is_iterable($object)) {
+            $stream = IterableStream::create($object);
+        } else {
+            throw new InvalidArgument('Third argument of SimpleS3Client::upload() must be string, resource, callable or an interable.');
+        }
+
+        return FixedSizeStream::create($stream, $chunkSize);
```

Comment:
```suggestion
        return new FixedSizeStream(StreamFactory::create($object), $chunkSize));
```

Reply:
Yes, But you are missing this line: 

`$stream = CallableStream::create($object, $chunkSize);`

Reply:
given the stream is wrapped in a `FixedStream` it will not be an issue. 
Maybe a performance issue, because the caller will try chunking the stream too much for nothing. 

My concernes is about dupplicate code (that does 99% the same thing) and maintanability (if we add a new Stream implementation, we'll have to update this code snippet at several places..

Maybe we should add a `int $preferedSize` in `Factory::create` to solve this

Reply:
Added